### PR TITLE
Terra 400 list implementation

### PR DESCRIPTION
**Pull request by dkasper-was-taken**
### Summary
The structural component to organize react elements within a list, as well as the selection styling for single selection lists and multi-selection lists.

[License]: ../blob/master/LICENSE

https://github.com/cerner/terra-core/issues/61

### CODE REVIEW

File: packages/terra-list/package.json [hunk 1]
```diff
@@ -0,0 +1,60 @@
+{
+  "name": "terra-list",
+  "main": "lib/List.js",
+  "version": "0.1.0",
+  "description": "terra-list",
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/cerner/terra-ui.git"
```
Comment (Matt-Butler) [inline packages/terra-list/package.json]: `terra-ui` => `terra-core`

File: jestconfig.json [hunk 2]
```diff
@@ -6,6 +6,6 @@
     "../../node_modules/enzyme-to-json/serializer"
   ],
   "moduleNameMapper": {
-    "\\.(css|scss)$": "identity-obj-proxy"
+    "\\.(css|scss|svg)$": "identity-obj-proxy"
```
Comment (Matt-Butler) [inline jestconfig.json]: Do we need this anymore? I don't think we are mocking any svgs

File: package.json [hunk 3]
```diff
@@ -30,7 +30,7 @@
       "<rootDir>/node_modules/enzyme-to-json/serializer"
     ],
     "moduleNameMapper": {
-      "\\.(css|scss)$": "identity-obj-proxy"
+      "\\.(css|scss|svg)$": "identity-obj-proxy"
```
Comment (Matt-Butler) [inline package.json]: same here

File: packages/terra-list/README.md [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+[![NPM version](http://img.shields.io/npm/v/terra-list.svg)](https://www.npmjs.org/package/terra-list)
+[![Build Status](https://travis-ci.org/cerner/terra-list.svg?branch=master)](https://travis-ci.org/cerner/terra-list)
+
+The Terra List is a structural component to arrange content within list/listitems.
+
+- [Getting Started](#getting-started)
+- [Documentation](#documentation)
+- [Supported Browsers](#supported-browsers)
+- [Contributing](#contributing)
+- [LICENSE](#license)
```
Comment (Matt-Butler) [inline packages/terra-list/README.md]: Can you please update this to match the structure of the other readmes?
```
- [Getting Started](#getting-started)
- [Documentation](https://github.com/cerner/terra-core/tree/master/packages/terra-list/docs)
- [LICENSE](#license)
```

File: packages/terra-list/README.md [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+## Getting Started
+
+- Install from [npmjs](https://www.npmjs.com)
+  - `npm install terra-list`
+  - `yarn install terra-list`
+- [Download the latest version](https://github.com/cerner/terra-list/archive/master.zip)
+- Clone the repo: `git clone https://github.com/cerner/terra-list.git`
```
Comment (Matt-Butler) [inline packages/terra-list/README.md]: remove line 20/21

File: packages/terra-list/README.md [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+## Documentation
```
Comment (Matt-Butler) [inline packages/terra-list/README.md]: Remove documentation and supported browsers

File: packages/terra-list/examples/MultiSelectListDemo.jsx [hunk 7]
```diff
@@ -0,0 +1,9 @@
+import React from 'react';
+
+import MultiSelectList from '../lib/MultiSelectList';
+
+export default () => <MultiSelectList isDivided maxSelectionCount={2}>
```
Comment (Matt-Butler) [inline packages/terra-list/examples/MultiSelectListDemo.jsx]: nit: can you please move exports to the bottom of these files to match the other components?

File: packages/terra-list/package.json [hunk 8]
```diff
@@ -0,0 +1,60 @@
+{
+  "name": "terra-list",
+  "main": "lib/List.js",
+  "version": "0.1.0",
```
Comment (Matt-Butler) [inline packages/terra-list/package.json]: "version": "0.0.0",

File: packages/terra-list/package.json [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+    "UI"
+  ],
+  "author": "Cerner Corporation",
+  "license": "Apache-2.0",
+  "bugs": {
+    "url": "https://github.com/cerner/terra-ui/issues"
+  },
+  "homepage": "https://github.com/cerner/terra-ui#readme",
+  "devDependencies": {
+    "csvtojson": "^1.1.4",
```
Comment (Matt-Butler) [inline packages/terra-list/package.json]: remove

File: packages/terra-list/package.json [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+    "jsdom": "^9.11.0",
```
Comment (Matt-Butler) [inline packages/terra-list/package.json]: remove

File: packages/terra-list/package.json [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+    "svgo": "^0.7.2",
```
Comment (Matt-Butler) [inline packages/terra-list/package.json]: remove

File: packages/terra-list/package.json [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+    "terra-markdown": "^0.1.0",
```
Comment (Matt-Butler) [inline packages/terra-list/package.json]: "terra-markdown": "^0.0.0",

File: packages/terra-list/package.json [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+    "terra-props-table": "^0.1.0",
```
Comment (Matt-Butler) [inline packages/terra-list/package.json]: "terra-props-table": "^0.0.0",

File: packages/terra-list/package.json [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+    "terra-toolkit": "^0.1.0"
+  },
+  "peerDependencies": {
+    "react": "^15.4.2",
+    "react-dom": "^15.4.2"
+  },
+  "dependencies": {
+    "classnames": "^2.2.5",
+    "terra-mixins": "^1.0.0",
+    "terra-icon": "^0.0.0"
```
Comment (Matt-Butler) [inline packages/terra-list/package.json]: terra-icon should probably be a devDependency.   Do we require icons in order to use list?

File: packages/terra-list/src/List.jsx [hunk 15]
```diff
@@ -0,0 +1,44 @@
+import React, { PropTypes } from 'react';
+import classNames from 'classnames';
+import '../src/List.scss';
```
Comment (Matt-Butler) [inline packages/terra-list/src/List.jsx]: import './List.scss';

File: packages/terra-list/src/ListItem.jsx [hunk 16]
```diff
@@ -0,0 +1,63 @@
+import React, { PropTypes } from 'react';
+import classNames from 'classnames';
+import ChevronRight from 'terra-icon/lib/icon/IconChevronRight';
+import '../src/ListItem.scss';
```
Comment (Matt-Butler) [inline packages/terra-list/src/ListItem.jsx]: `import '../ListItem.scss';`

File: packages/terra-list/README.md [hunk 17]
```diff
@@ -0,0 +1,52 @@
+# Terra List
+
+
+[![NPM version](http://img.shields.io/npm/v/terra-list.svg)](https://www.npmjs.org/package/terra-list)
+[![Build Status](https://travis-ci.org/cerner/terra-list.svg?branch=master)](https://travis-ci.org/cerner/terra-list)
```
Comment (Matt-Butler) [inline packages/terra-list/README.md]: [Build Status](https://travis-ci.org/cerner/terra-core)

File: packages/terra-list/lib/List.scss [hunk 18]
```diff
@@ -0,0 +1,15 @@
+@import '~terra-mixins';
+@import './variables';
+@import './mixins';
```
Comment (Matt-Butler) [inline packages/terra-list/lib/List.scss]: Can we remove the files for mixins  / variables?

File: packages/terra-list/lib/ListItem.scss [hunk 19]
```diff
@@ -0,0 +1,42 @@
...
+  display: block;
+  position: relative; // relative position so the chevron can use absolute position to vertically center itself
+
+  [dir='ltr'] & {
+    padding-right: 20px;
+  }
+
+  [dir='rtl'] & {
+    padding-left: 20px;
+  }
```
Comment (Matt-Butler) [inline packages/terra-list/lib/ListItem.scss]: We have a mixin for padding end:

```
@include terra-padding-end(20px);
```

**Comment by Matt-Butler** [comment]
For selectable lists, we need to add `tabindex="0"` to make them accessible.

**Comment by dkasper-was-taken** [comment]
Added tab index.

File: packages/terra-list/package.json (hunk 14 above)
Comment (dkasper-was-taken) [inline packages/terra-list/package.json]: Verified, we need the dependency for chevron.

File: packages/terra-list/docs/README.md [hunk 20]
```diff
@@ -0,0 +1,63 @@
...
+
+- Install with [npmjs](https://www.npmjs.com):
+  - `npm install terra-list`
+  - `yarn add terra-list`
+
+## Usage
+
+```jsx
+import React from 'react';
+import SlidePanel from 'terra-slide-panel';
```
Comment (Matt-Butler) [inline packages/terra-list/docs/README.md]: Update

**Comment by Matt-Butler** [comment]
multiselect
For the multi-select example, I am only able to select two list items. Is this the expected behavior?

File: packages/terra-list/src/MultiSelectList.jsx [hunk 21]
```diff
@@ -0,0 +1,180 @@
...
+  render() {
+    const { children, isDivided, onChange, maxSelectionCount, ...customProps } = this.props;
+    const clonedChildItems = this.cloneChildItems(children);
+
+    // Figure out how to handle this scenario.
+    this.unusedVariables(onChange);
+    this.unusedVariables(maxSelectionCount);
+
+    return (
+      <List isDivided={isDivided} {...customProps} tabindex="0">
```
Comment (Matt-Butler) [inline packages/terra-list/src/MultiSelectList.jsx]: We need to camelCase tabindex
`tabindex` => `tabIndex`

In react, almost all hyphenated html attributes need to be camelcased: https://facebook.github.io/react/docs/dom-elements.html#all-supported-html-attributes

**Comment by dkasper-was-taken** [comment]
In the example I set a max count of 2, but default is no maximum.

**Pull request merged**